### PR TITLE
Add Basic Matrix Wrapper

### PR DESCRIPTION
**Pull request by isaaclegred**
### Types of changes:

- [ ] Bugfix
- [x ] New feature

### Component:

- [ x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/DataStructures/Python/CMakeLists.txt
Comment (nilsdeppe) [inline src/DataStructures/Python/CMakeLists.txt]: Could you change this to now be:
```cmake
  SOURCES
  Bindings.cpp
  DataVector.cpp
  Matrix.cpp
```
? The reason is this way if two people add or modify the list, it's less likely to result in a conflict :)

File: src/DataStructures/Python/DataVector.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/DataVector.cpp]: Oops, this is not the formatting we want :)

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: Is this a `TODO`? If so, you should use `TODO` in the code so we know it's a TODO and should be resolved before merging :)

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: This can just be `return std::string(MakeString{} << M);`.

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: `M` is not snake_case and not descriptive. maybe something like `matrix` would be better, or `self`

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: just do `return bp::make_tuple...` (also your formatting needs work :) ) I think you can run `git clang-format` and `clang-format` on Wheeler

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: please remove comment

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: Can you please remove all the extra blank lines here?

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp:16]: please add a blank line before the `namespace bp = ` so it's clearly separate from the includes

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp:4]: Can you sort the includes according to our guidelines please?

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp:44]: missing closing namespace comment

File: tests/Unit/DataStructures/Test_Matrix.py
Comment (nilsdeppe) [inline tests/Unit/DataStructures/Test_Matrix.py]: This looks to be over our 80 character limit

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [inline src/DataStructures/Python/Matrix.cpp]: please run clang-format over this file

**Review by nilsdeppe — DISMISSED** [review]
I'm happy for you to squash your fixups and rebase this on develop

**Comment by nilsdeppe** [comment]
@nilsleiffischer can you please review this?

File: src/DataStructures/Python/CMakeLists.txt
Comment (nilsvu) [inline src/DataStructures/Python/CMakeLists.txt]: remove space

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsvu) [inline src/DataStructures/Python/Matrix.cpp]: To be consistent with Python conventions, I think this should be an attribute named `shape` returning a `tuple`:

```python
In[0]: Matrix(3, 4).shape
Out[0]: (3, 4)
```

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsvu) [inline src/DataStructures/Python/Matrix.cpp]: What's the purpose of this comment? Remove or rephrase to make more clear what you want to tell the reader

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsvu) [inline src/DataStructures/Python/Matrix.cpp:21]: We'll also need an initialiser from a nested list I think:

```python
Matrix([[1., 2., 3.], [4., 5., 6.]])
```

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsvu) [inline src/DataStructures/Python/Matrix.cpp]: should this be `x[1]`?

Fix, then add a test that would have caught this

**Review by nilsvu — CHANGES_REQUESTED** [review]
Also @nilsdeppe:

`Matrix` ist just a specialisation of `DenseMatrix` that fixes the storage order to column-major. I guess we only need this thin wrapper for `Matrix` for now to work with cpp code that uses it? So also wrapping the math operations provided by blaze may not be necessary for now?

We should make sure that at least a `Matrix` can be converted to a `np.ndarray`, so please add a test for that:

```python
np.testing.assert_equal(np.array(Matrix([[1., 2.], [3., 4.]])), np.array([[1., 2.], [3., 4.]]))
```

**Comment by nilsdeppe** [comment]
The plan is to add the conversion to np arrays in a followup PR. Is that okay with you?

**Comment by nilsvu** [comment]
Sure, it‘s not urgent. @isaaclegred please open an issue for that.

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsdeppe) [reply src/DataStructures/Python/Matrix.cpp:21]: Nope, we're not planning on using Matrix is this way. It's a temporary class we need for loading Dat files. Ultimately we want the Matrix class to be converted directly to a numpy matrix. It won't ever be constructed in Python code, only in C++ code (read-only object)

File: src/DataStructures/Python/Matrix.cpp
Comment (nilsvu) [inline src/DataStructures/Python/Matrix.cpp]: make this an attribute please (`add_property` in boost)

File: tests/Unit/DataStructures/Test_Matrix.py
Comment (nilsvu) [inline tests/Unit/DataStructures/Test_Matrix.py]: ... this should be `M.shape` then

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/DataStructures/Python/Matrix.cpp
Comment (isaaclegred) [reply src/DataStructures/Python/Matrix.cpp]: This is pretty difficult to do, since shape is not a member of the Dynamic Matrix class, and there doesn't seem to be any shape function for the Blaze Matrices.  I could make separate attributes for rows and columns but otherwise I don't know how to make this feasible.

Comment (nilsdeppe) [reply src/DataStructures/Python/Matrix.cpp]: I don't know how to do this either. Given that this class is temporary for getting the rest of the IO parts wrapped I'd prefer not to hold up on this. Would that be okay with you @nilsleiffischer? The long term plan is to figure out how to turn `Matrix` into a numpy array anyway.

Comment (nilsvu) [reply src/DataStructures/Python/Matrix.cpp]: It should be a simple matter of passing `add_property` a getter function, see https://www.boost.org/doc/libs/1_61_0/libs/python/doc/html/tutorial/tutorial/exposing.html#tutorial.exposing.class_properties
If this turns out too complicated, we can postpone, but it appears quite simple on first glance.

Comment (nilsvu) [reply src/DataStructures/Python/Matrix.cpp]: In fact (still on first glance) it looks to me this would work by just replacing `def` with `add_property`.

Comment (isaaclegred) [reply src/DataStructures/Python/Matrix.cpp]: Yeah, sorry, I guess that does work,  thanks, I didn't realize the getter function could take an argument.

**Comment by nilsdeppe** [comment]
@isaaclegred if you have time could you squash and rebase this on the latest develop?

@nilsleiffischer could you please look at the update?

**Review by nilsvu — DISMISSED** [review]

**Review by nilsdeppe — APPROVED** [review]
@kidder please look at this. It should be ready to go.

**Review by kidder — APPROVED** [review]

**Pull request merged**
